### ISSUE TITLE
[Bug] Gradient computation issue in methods/dps.py

### ISSUE THREAD

**Issue by TianyiLi-92** [YWolfeee/Training-Free-Guidance#5]
# [Bug] Gradient computation issue in `methods/dps.py`

Thank you for this excellent work! It has been very helpful in my research.  
While integrating `methods/dps.py` into my codebase, I noticed a potential logic error related to gradient computation.
Could you please confirm if my understanding is correct? Let me know if this suggestion aligns with your intended design.

## Problem Description
In `methods/dps.py`, lines 30–35:
```python
func = lambda zt: (zt - (1 - alpha_prod_t) ** (0.5) * unet(zt, t)) / (alpha_prod_t ** (0.5))

guidance = self.guider.get_guidance(x.clone().detach().requires_grad_(), func, **kwargs)
        
# follow the schedule of DPS paper
logp_norm = self.guider.get_guidance(x.clone().detach(), func, return_logp=True, check_grad=False, **kwargs)
```

Currently, the gradient computation in the following section appears to take the gradient with respect to the **updated `x_need_grad`** (which is an estimation of `x0`) rather than the original `xt`. This occurs because `x_need_grad` is reassigned before computing the gradient. For example, in `tasks/audio_inpainting.py`:

```python
x_need_grad = func(x_need_grad)  # Line 69
grad = torch.autograd.grad(log_probs.sum(), x_need_grad)[0]  # Line 93
```

This results in `torch.autograd.grad` incorrectly computing gradients based on the modified `x_need_grad`.

## Suggested Fix
In `methods/dps.py`, replace lines 30–35 with:

```python
x = x.clone().detach().requires_grad_()

x0 = (x - (1 - alpha_prod_t) ** (0.5) * unet(x, t)) / (alpha_prod_t ** (0.5))

logp_norm = self.guider.get_guidance(x0, return_logp=True, check_grad=False, **kwargs)

guidance = torch.autograd.grad(logp_norm.sum(), x)[0]
```

This ensures the gradient computation is correctly performed on `x` (noisy images/audios) rather than the estimated `x0`.

**Comment by YWolfeee** [comment]
Thanks for bringing this up, and indeed, the reassignment becomes a problem for gradient computation. Fortunately, we double checked the code and this only exists in our newly prepared codebase (so it doesn't influence our paper's results).

As a solution, we plan to replace
```
x_need_grad = func(x_need_grad)

x = post_process(x_need_grad)
```
into
```
x = post_process(func(x_need_grad))
```

This solution automatically applies to all methods (beyond DPS). Please let us know if that sounds good to you.

**Comment by TianyiLi-92** [comment]
Thanks! Just wanted to check my understanding. The solution seems great. Really appreciate your prompt response and careful verification!
